core/src/main/java/org/apache/calcite/rel/metadata/RelMetadataQuery.java

```diff
@@ -75,8 +76,17 @@
  * plugin mechanism.
  */
 public class RelMetadataQuery {
-  /** Set of active metadata queries, and cache of previous results. */
-  public final Map<List, Object> map = new HashMap<>();
+  /**
+   * Set of active metadata queries, and cache of previous results.
+   *
+   * Can be sealed when no more additions to the map are expected.
+   * Write operations on the map will fail after sealing
+   *
+   * Sealing is helpful if no additions to the map are expected or desired.
+   * The map here is unbounded so it can leak memory if not properly cleaned
+   * up
+   */
+  public final SealableMap<List, Object> map = new SealableMap<>(new HashMap<>());
```

Comment:
Just make it `public final Map<List, Object> map = new SealableMap<>(new HashMap<>());`

Reply:
It needs to be a `SealableMap` because the APIs of `SealableMap` need to be used by whoever wants to seal this map. O/w they would have to know its a `SealableMap` and cast it